### ISSUE TITLE
Regression: DDL can't run inside a transaction block

### ISSUE THREAD

**Issue by lfittl** [citusdata/citus#668]
On current master:

```
postgres=# BEGIN;                                  
BEGIN
postgres=# CREATE TABLE test (id uuid PRIMARY KEY);
CREATE TABLE
postgres=# SELECT master_create_distributed_table('test', 'id', 'hash');
 master_create_distributed_table 
---------------------------------
 
(1 row)

postgres=# SELECT master_create_worker_shards('test', 16, 1);
 master_create_worker_shards 
-----------------------------
 
(1 row)

postgres=# ALTER TABLE test ADD COLUMN description text;
ERROR:  distributed DDL commands cannot run inside a transaction block
postgres=# ROLLBACK;
ROLLBACK
```

On 5.1:

```
citus=> BEGIN;
BEGIN
citus=> CREATE TABLE test (id uuid PRIMARY KEY);
CREATE TABLE
citus=> SELECT master_create_distributed_table('test', 'id', 'hash');
 master_create_distributed_table 
---------------------------------
 
(1 row)

citus=> SELECT master_create_worker_shards('test', 16, 1);
 master_create_worker_shards 
-----------------------------
 
(1 row)

citus=> ALTER TABLE test ADD COLUMN description text;
ALTER TABLE
citus=> COMMIT;
COMMIT
```

This is a regression on master, and introduces a new problem for e.g. Rails, since it runs all migrations (and the associated DDLs) in a wrapping transaction. You can disable DDL transactions in Rails, but then you end up with half-done migrations (which Rails does not keep track off accurately).

Sidenote: We also seem to be leaking the individual shard tables when a ROLLBACK is issued - we should probably use 2PC for `master_create_worker_shards` now that we have it.

cc @jasonmp85 (via https://github.com/citusdata/citus/issues/666)

# Implementation Plan

This regression will be addressed in two phases…

## Phase One

Re-enables blocks with single DDL statements while putting in place safeguards to prevent interlacing of DDL and DML statements within a single block

**Completion Estimate**: Around a day

### Implementation

Around a day to write, review, and test

  - [x] Add bitfield to track types of operations (DDL, DML, others?) yet performed in an open transaction block
  - [x] Add `|=` assignments to paths that perform these operations, to note them in the field
  - [x] Before DML commands, check this field to ensure at most the DML flag is set (error otherwise)
  - [x] Before DDL commands, check this field and error if any fields are set (error otherwise)
  - [x] Add explicit tests for the above two items

## Phase Two

Expands DDL block handling to permit multiple DDL statements within a single block

**Completion Estimate**: Three days

### Implementation

Less than a day to write and review; one or two days to hand-test

  - [x] Remove DDL bit field check within DDL handling code (still error if DML field set)
  - [x] Re-use list of placement connections to perform subsequent DDL commands
  - [x] Add tests for multi DDL blocks (happy case)
  - [x] Add tests for multi DDL blocks (`ROLLBACK` happy case)
  - [x] Add tests for multi DDL blocks (partial failures)

The reviewer should download and compile the second phase's branch in order to perform manual verification of the changes.

**Comment by marcocitus** [comment]
DDL in transaction blocks was always dangerously broken to the point where you really shouldn't be using it, so it's more of a clarification than a regression right now. We're actually not far off from supporting DDL in transaction blocks and it might be useful to go all the way. The main concern is how it interacts with other things you can do in transaction blocks such as DML. It might make sense to allow either DML or DDL in a transaction, but not both.

**Comment by ozgune** [comment]
@marcocitus -- I had a clarification question. Where were DDLs in transactions were dangerously broken? Could you elaborate a bit more on our previous behavior?

**Comment by anarazel** [comment]
On 2016-07-25 10:16:55 -0700, Ozgun Erdogan wrote:
> @marcocitus -- I had a clarification question. Where were DDLs in transactions were dangerously broken? Could you elaborate a bit more on our previous behavior?

BEGIN;
ALTER TABLE;
ROLLBACK;

left the workers in a different state than the master. But the fix for
that only requires using the transaciton hook, not necessarily 2pc.

**Comment by ozgune** [comment]
This is a usability regression from our current behavior for say, Rails. Lukas noted that changing Ruby config for schema migrations is something he could live with for v5.2. Andres also noted that the implications of our current transaction support may be bigger than this issue.

We internally synced up on this. @marcocitus noted that the reason we error out here is that: if the user starts a transaction, runs ``INSERT`` and ``DDL``, we end up with a distributed deadlock. Similarly, if the user starts a transaction, runs ``DDL`` and ``INSERT``, we deadlock.

The underlying reason for just erroring out here is that we had two concurrent pull requests, DDL propagation and transaction support, that made it to master near the end of our release timeline.

@marcocitus proposed a simple fix where we have modes (DDL, DML, etc.) and only error out if the user runs transactions and issues commands that have mixed modes. Time estimates were half a day for the fix and 2 days for fixing.

@marcocitus -- if you have additional notes on this (for example, IsTopLevel(), etc.), could you add them as comments to this issue?

**Comment by jasonmp85** [comment]
Now that I've got a handle on how the existing DDL code works (I believed it still did not use transaction hooks), I have a proposal for how to tackle this. I'm putting it up top with a bit of implementation details.

Essentially, there will be two (serial) phases (i.e. two pull requests) to restore the pre-5.2 user experience.

After the first PR is merged, users will be able to run a DDL command inside of a transaction block, so long as they do not attempt to run a _second_ DDL command (which will throw an error).

Attempting to run a DDL command in a block which already has performed modifications, or attempting to run modifications in a block which has already performed DDL command is likewise an error.

This first change is low-impact as it doesn't really alter the existing behavior: we already use commit callbacks to finish DDL processing, and we already don't allow interleaving of DDL and DML commands within a transaction block (because we don't allow DDL in a transaction block whatsoever right now), so updating regression tests should be sufficient for the testing process (in addition to the usual code review).

The second change will re-enable the ability to use multiple DDL statements within a single transaction block, except now (unlike in 5.1), they will actually be performed atomically. As before (and due to the changes made in the first phase), it will remain impossible to interlace DDL and DML commands, but multi-statement blocks purely consisting of one or the other will be honored (and atomic).

Though this second change will be less code, it will require more testing, as we have—heretofore—not supported atomic blocks of multiple DDL statements.

**Comment by samay-sharma** [comment]
Feedback from a user: "Our drop table commands in our migration scripts do not work inside of a transaction scope, which makes perfect sense why it’s not supported.  However we have no control over our migration tool, and all DB migration frameworks and ORMs I’ve ever used wrap migrations inside of transactions always.  So it would be nice if we could “turn off” Citus’ protection layer here and proceed with the distributed DDL even if we’re breaking out of the transaction’s semantics."

**Comment by marcocitus** [comment]
@samay-sharma DROP table follows an entirely different code path than other DDL statements. It might be worth opening a separate issue. So we can close this one once #764 is merged.

**Comment by samay-sharma** [comment]
@marcocitus : Ah, right. Opened #774 , and copied the comment there.

**Comment by jasonmp85** [comment]
Oops, meant to close this with the merge. Closing now, complete.
